Move dashboard card action buttons in a popover

alxnddr opened the pull request:
## Description

### 1. Changing text cards and card actions UI

#### Before
Currently, 1 and 2 rows height text cards look not great in editing dashboard mode:
<img width="326" alt="Screen Shot 2021-04-15 at 23 20 52" src="https://user-images.githubusercontent.com/14301985/114933481-62a55f80-9e41-11eb-9b07-487bb55e272a.png">
<img width="325" alt="Screen Shot 2021-04-15 at 23 20 57" src="https://user-images.githubusercontent.com/14301985/114933493-6507b980-9e41-11eb-8fe4-9384b56245c1.png">
In addition to that, we decrease height with decreasing width of a window which leads to:
<img width="198" alt="Screen Shot 2021-04-15 at 23 22 36" src="https://user-images.githubusercontent.com/14301985/114933679-9ed8c000-9e41-11eb-8d45-2ae2787c9e20.png">

#### After
In order to keep 1-row height text cards usable, we decided not to shrink height on narrow screens and to move action buttons out of the card and show them on hover:
<img width="324" alt="Screen Shot 2021-04-19 at 23 54 18" src="https://user-images.githubusercontent.com/14301985/115302302-f725ff80-a16a-11eb-9cc9-2b8ac7db2834.png">

In order to make UX consistent, it was applied to all dashboard cards:
<img width="787" alt="Screen Shot 2021-04-19 at 23 55 01" src="https://user-images.githubusercontent.com/14301985/115302335-0147fe00-a16b-11eb-9e12-7452ae3f1e32.png">

It prevents having an overcrowded header on small cards with charts (before):
<img width="272" alt="Screen Shot 2021-04-22 at 19 55 23" src="https://user-images.githubusercontent.com/14301985/115754251-ae598b00-a3a4-11eb-89a8-dbd6395c2c74.png">

### 2. Set min height for rows of the dashboard grid

Set reasonable min height for a row so that cards on narrow screens are not that squeezed

#### Before
<img width="454" alt="Screen Shot 2021-04-20 at 00 00 18" src="https://user-images.githubusercontent.com/14301985/115303244-0c4f5e00-a16c-11eb-8419-9a0d611d1f4e.png">

#### After
<img width="457" alt="Screen Shot 2021-04-20 at 00 00 24" src="https://user-images.githubusercontent.com/14301985/115303365-30ab3a80-a16c-11eb-9288-3fd6fcd22aaf.png">


### 3. Set min height for text cards for mobile screen resolutions

#### Before
<img width="357" alt="Screen Shot 2021-04-20 at 00 01 45" src="https://user-images.githubusercontent.com/14301985/115303404-3d2f9300-a16c-11eb-8915-75492bcdaca9.png">

#### After
<img width="371" alt="Screen Shot 2021-04-20 at 00 01 50" src="https://user-images.githubusercontent.com/14301985/115303417-428cdd80-a16c-11eb-854e-d10624d9f18d.png">

alxnddr committed:
move dashboard card action buttons in a popover

alxnddr committed:
add more top margin for actions menu when editing dashboard, fix tests

ukritw commented:
On paper, this looks nice!

I have two questions though:
* Where would the user need to hover to see the action card? 
    * imo the gesture to get to action card should be a bit more forgiving than what it is right now if the users have to move the cursor above the cards
* What would happen if the action card overlap on a question card above? 
<img width="361" alt="Screen Shot 2021-04-16 at 2 03 03 PM" src="https://user-images.githubusercontent.com/2685753/114985826-75f61080-9ebd-11eb-8143-ea97f18627e1.png">

alxnddr commented:
@ukritw thanks for the feedback!
1) Users need to hover anywhere on the card except the resize handle in the right bottom corner
2) These action cards are above dashboard cards, I recorded a [quick video to show this](https://share.getcloudapp.com/L1udd0OE)

ukritw commented:
Cool, that's perfect then.

mazameli commented:
I have some pretty major reservations about this. I think we might be degrading the UX for the majority of cases for the sake of solving the small text card problem. Have we considered doing this _only_ for small text cards? In other words, would we being doing this if text cards didn't exist? I'm not sure I see the benefit for other cards, and I'm not personally worried about small text cards having a slightly different interaction when needed.

**Some specific reservations:**

Funky way the controls overlap on other cards:

![image](https://user-images.githubusercontent.com/2223916/115059404-5ea62b80-9e9b-11eb-8701-84a513bb6875.png)

Requires finer mouse control to keep the options visible:

![Screen+Recording+2021-04-16+at+10 07 06+AM](https://user-images.githubusercontent.com/2223916/115059602-9e6d1300-9e9b-11eb-8fc4-849b68e711db.gif)

nemanjaglumac commented:
This pattern repeats enough that it makes sense to extract it into a custom function or Cypress custom command:
```javascript
cy.get(".DashCard").trigger("mouseover");
cy.icon("foo").click({ force: true });
```
Don't have to do it in this PR  - this was more a general comment for the future reference.

nemanjaglumac commented:
When cards are stacked on top of each other, there is a little bit of overlap. Not a big deal, imo but I thought it's worth bringing it up just in case. cc @mazameli 
![image](https://user-images.githubusercontent.com/31325167/115396164-fcc02b80-a1e4-11eb-9ac1-7b5d7db7b43c.png)

Increasing the margin between cards to 20px solves that, but then cards have a bit more room when not in editing mode.
![image](https://user-images.githubusercontent.com/31325167/115396469-4c9ef280-a1e5-11eb-8cbe-fe11afc2faf9.png)

kdoh commented:
@nemanjaglumac Yeah that overlap is to be expected. The idea is to try and work within the space we already have while providing relief from crowding in the card headers, so the floating actions being slightly offset is a way to accomplish that while still retaining a relationship to the card you're hovering on.

nemanjaglumac commented:
There is a scrollbar on one line text box in full-screen mode. Don't take my word for it, but I _think_ we solved it before. Not sure if changes here have anything to do with it, tho.
![image](https://user-images.githubusercontent.com/31325167/115403290-90e1c100-a1ec-11eb-9a68-e1957b09f849.png)

Dark mode doesn't influence this - chose it simply because it's easier to see the scrollbar.

kdoh reviewed (approved):
Looks good!

daltojohnso commented on frontend/src/metabase/dashboard/components/DashCard.jsx:
```diff
@@ -144,6 +159,10 @@ export default class DashCard extends Component {
 
+    const isEditingDashboardLayout =
+      isEditing &&
+      !(clickBehaviorSidebarDashcard != null || isEditingParameter);
```
I think the double negation here (the `!(foo != null)` part) is a little confusing. Can you change it to this:

```
    const isEditingDashboardLayout = isEditing &&
        clickBehaviorSidebarDashcard == null &&
        !isEditingParameter;
```

That way it sorta reads, "`isEditingDashboardLayout` is `true` when we're editing, there's no `clickBehaviorSidebarDashcard`, and I'm not editing a parameter"

alxnddr replied:
I agree that it reads easier this way, thanks!

daltojohnso reviewed (changes requested):
one nitpick, but looks good

alxnddr committed:
Address review comments

alxnddr commented:
@nemanjaglumac thanks for the review, I addressed all comments.
Since not a while ago I started using css `:hover` to show the menu, triggering `mouseover` did not help at all. To simulate css hover I added `cypress-real-events` package and used `realHover` command and extracted it into a separate function.

Regarding text cards in full-screen mode, the problem existed before but I fixed it by not increasing the font in fullscreen mode only for single-row text cards.

alxnddr committed:
Fix spec broken due to changed height of dashboard cards
